kani-compiler/src/kani_middle/stubbing/transform.rs

```diff
@@ -23,18 +26,57 @@ pub fn get_stub(tcx: TyCtxt, def_id: DefId) -> Option<DefId> {
 
 /// Returns the new body of a function/method if it has been stubbed out;
 /// otherwise, returns the old body.
-pub fn transform<'tcx>(
-    tcx: TyCtxt<'tcx>,
-    def_id: DefId,
-    old_body: &'tcx Body<'tcx>,
-) -> &'tcx Body<'tcx> {
+pub fn transform<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, old_body: &'tcx Body<'tcx>) -> Body<'tcx> {
     if let Some(replacement) = get_stub(tcx, def_id) {
         let new_body = tcx.optimized_mir(replacement).clone();
         if check_compatibility(tcx, def_id, old_body, replacement, &new_body) {
-            return tcx.arena.alloc(new_body);
+            return new_body;
+        }
+    }
+    old_body.clone()
+}
+
+/// Tranverse `body` searching for calls to foreing functions and, whevever there is
+/// a stub available, replace the call to the foreign function with a call
+/// to its correspondent stub. This happens as a separate step because there is no
+/// body available to foreign functions at this stage.
+pub fn transform_foreign_functions<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    let mut visitor = if let Some(stub_map) = get_stub_mapping(tcx) {
+        ForeignFunctionTransformer { tcx, body: body.clone(), stub_map }
+    } else {
+        ForeignFunctionTransformer { tcx, body: body.clone(), stub_map: HashMap::default() }
+    };
+    visitor.visit_body(body);
+}
+
+struct ForeignFunctionTransformer<'tcx> {
+    /// The compiler context.
+    tcx: TyCtxt<'tcx>,
+    /// Body of the callee function. Kani searches here for foreign functions.
+    body: Body<'tcx>,
+    /// Map of functions/methods to their correspondent stubs.
+    stub_map: HashMap<DefId, DefId>,
+}
+
+impl<'tcx> MutVisitor<'tcx> for ForeignFunctionTransformer<'tcx> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_operand(&mut self, operand: &mut Operand<'tcx>, _location: Location) {
+        let func_ty = operand.ty(&self.body, self.tcx);
+        if let ty::FnDef(reachable_function, generics) = *func_ty.kind() {
+            if self.tcx.is_foreign_item(reachable_function) {
+                if let Some(stub) = self.stub_map.get(&reachable_function) {
+                    let Operand::Constant(fn_def) = operand else { unreachable!() };
```

Comment:
I still think this would get triggered if you do something like this
```rs
let ptr = function_name;
ptr(input)
```

What I mean to say is that I think this doesn't get triggered in your test case because the type of `ptr` is something generic you set it to but in the example I provided the type would be exactly type of `function_name` and trigger this code path. 
I don't think there is any expectation that this is a common path, but there is also no reason why we shouldn't handle that case.

I'd propose we add another test case where the function pointer variable is not a generic type and see if it crashes. It might be that constant propagation actually takes care of this. 
If it does crash you could simply do `return` instead of `unreachable!()` here and everything should work just fine.

Reply:
I added a new test case without generic type and replaced `unreachable!()` with `return`, even tho it made no difference in the test case.

Reply:
Great. Then my suspicions were unwarranted it seems. Now we know :)